sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpSystemProperties.cs

```diff
@@ -0,0 +1,380 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using Azure.Core.Amqp;
+using Microsoft.Azure.Amqp.Framing;
+
+namespace Azure.Messaging.EventHubs.Amqp
+{
+    /// <summary>
+    ///   Provides an abstraction over an <see cref="AmqpAnnotatedMessage" /> to represent
+    ///   broker-owned data in the form of a "system properties" dictionary for the
+    ///   associated <see cref="EventData" /> instance.
+    /// </summary>
+    ///
+    /// <remarks>
+    ///   This wrapper is necessary to preserve the behavior of the <see cref="EventData.SystemProperties" />
+    ///   member without potentially drifting from the mutable data that exists on the AMQP message.
+    /// </remarks>
+    ///
+    /// <seealso cref="IReadOnlyDictionary{TKey, TValue}" />
+    ///
+    internal class AmqpSystemProperties : IReadOnlyDictionary<string, object>
+    {
+        /// <summary>The set of system properties that are sourced from the Properties section of the <see cref="AmqpAnnotatedMessage" />.</summary>
+        private static readonly string[] PropertySectionNames = new[]
+        {
+           Properties.MessageIdName,
+           Properties.UserIdName,
+           Properties.ToName,
+           Properties.SubjectName,
+           Properties.ReplyToName,
+           Properties.CorrelationIdName,
+           Properties.ContentTypeName,
+           Properties.ContentEncodingName,
+           Properties.AbsoluteExpiryTimeName,
+           Properties.CreationTimeName,
+           Properties.GroupIdName,
+           Properties.GroupSequenceName,
+           Properties.ReplyToGroupIdName
+        };
+
+        /// <summary>The AMQP message to use as the source for the system properties data.</summary>
+        private readonly AmqpAnnotatedMessage _amqpMessage;
+
+        /// <summary>
+        ///   Attempts to retrieve the value of the requested <paramref name="key" />
+        ///   from the dictionary.
+        /// </summary>
+        ///
+        /// <param name="key">The key of the dictionary item to retrieve.</param>
+        ///
+        /// <returns>The value in the dictionary associated with the specified <paramref name="key"/>.</returns>
+        ///
+        /// <exception cref="KeyNotFoundException">Occurs when the specified <paramref name="key" /> does not exist in the dictionary.</exception>
+        ///
+        public object this[string key]
```

Comment:
Is perf a concern at all for this type?

Reply:
Yes and no;  the callers for `SystemProperties` aren't many and won't be calling in a tight loop.   That said, this isn't optimal with respect to performance for callers who are probing the dictionary frequently.   Unfortunately, with `AmqpAnnotatedMessage` being fully mutable and offering nothing like `INotifyPropertyChanged` to detect updates,  we can't safely cache values here or even generate an actual dictionary with copied data.

If you've got an idea, I'm all ears.   This was the best trade-off that I could come up with for safety.

Reply:
Not sure if this is better without profiling, but I was thinking something like a `Dictionary<string, Func<AmqpAnnotatedMessage, object>>`. Then an example get below would look like:

```c#
return _dict[key](_amqpMessage);
```

Reply:
The `SystemProperties` dictionary on `EventData` is part of the GA surface, so it has to remain `IReadOnlyDictionary<string, object>` with respect to the public interface.  In theory, we could wrap your proposed implementation.

That said, there's a couple of considerations that I see:

- Because the data comes from mutable data in the underlying AMQP message, we don't have a stable set of keys to pre-populate in the key-to-func dictionary.   The well-known properties would work as accessors, but we'd need to do the same condition check to determine a response for `ContainsKey` since we don't want the key to be present if the section or value is not populated.  _(This is to preserve existing behavior, there are no null values ever added to the system properties set in GA)._   

- Since we can't capture the set of annotations and create keys for them in the func dictionary, because items may be added/removed from the underlying set, we'd have to find a way to intercept any unrecognized key and dynamically map to the annotations section.

- Because each section of the AMQP message does lazy allocation when you use the accessor, we want to continue to guard with `HasSection` before accessing the property directly, so that we don't trigger unnecessary allocations.   We could do this in the func, but it prevents a simple pass-through.

- I see us having to replicate the majority of logic from the custom dictionary projection, since we can't assume a stable set of keys to prepopulate the func dictionary and we can't cache;  I worry that we'd end up allocating a set of func instances that just delegated to something similar to this form.  I don't see us saving much in performance since we'd have to run much of the same logic, and I see us eating more allocations in that scenario.

It could be that I'm overlooking some of your thoughts.   If you're game, I'd like to move forward with the current form to not lose momentum and continue talking over your ideas to potentially follow in another set of changes.